File: frontend/src/types/v2Report.ts
export interface V2ReportFilters {
  date_from?: string;
  date_until?: string;
  category_id?: string;
  responsible_user_id?: string;
}

export interface V2ReportSummary {
  local_date: string;
  date_from: string | null;
  date_until: string | null;
  category_id: string | null;
  responsible_user_id: string | null;
  counts: {
    tasks: number;
    pending_items: number;
    projects: number;
    activities: number;
    total: number;
  };
}

export interface TaskReportMetrics { total_count: number; pending_count: number; completed_count: number; not_completed_count: number; resolved_count: number; completion_rate: string | null }
export interface TaskReportGroup extends TaskReportMetrics { key: string; label: string }
export interface TaskReportEvolution extends TaskReportMetrics { planned_date: string }
export interface V2TaskReport { period: { date_from: string | null; date_until: string | null }; filters: V2ReportFilters; master_task_id: string | null; custom_tasks: boolean | null; summary: TaskReportMetrics; by_task: TaskReportGroup[]; by_category: TaskReportGroup[]; evolution: TaskReportEvolution[] }
export interface ProgressReportMetrics { total_count: number; no_iniciado_count: number; en_proceso_count: number; finalizado_count: number; configuracion_incompleta_count?: number; average_progress: string | null }
export interface ComplianceMetrics { en_plazo_count: number; atrasado_count: number; con_adelanto_count: number; a_tiempo_count: number; con_retraso_count: number }
export interface ProgressCategoryGroup extends ProgressReportMetrics { category_id: string; category_name: string }
export interface ProgressEvolution { planned_date: string; total_count: number; average_progress: string | null }
export interface V2PendingReport { period: { date_from: string | null; date_until: string | null }; filters: V2ReportFilters; summary: ProgressReportMetrics; compliance: ComplianceMetrics; by_category: ProgressCategoryGroup[]; evolution: ProgressEvolution[] }
export interface ProjectReportRow { project_id: string; project_name: string; category_id: string; category_name: string; planned_date: string | null; progress: string | null; state: string; stage_count: number }
export interface V2ProjectReport { period: { date_from: string | null; date_until: string | null }; filters: V2ReportFilters; summary: ProgressReportMetrics; stage_compliance: ComplianceMetrics; by_category: ProgressCategoryGroup[]; by_project: ProjectReportRow[]; evolution: ProgressEvolution[] }
export interface ActivityReportMetrics { total_count: number; scheduled_count: number; cancelled_count: number; total_duration_minutes: string; average_duration_minutes: string | null }
export interface ActivityReportGroup extends ActivityReportMetrics { key: string; label: string }
export interface ActivityReportEvolution extends ActivityReportMetrics { local_date: string }
export interface V2ActivityReport { period: { date_from: string | null; date_until: string | null }; filters: V2ReportFilters; activity_master_id: string | null; custom_activities: boolean | null; summary: ActivityReportMetrics; by_activity: ActivityReportGroup[]; by_category: ActivityReportGroup[]; by_organizer: ActivityReportGroup[]; evolution: ActivityReportEvolution[] }

File: backend/app/schemas/v2_report.py
import uuid

from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class _StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class ReportSummaryCounts(_StrictModel):
    tasks: int = Field(ge=0)
    pending_items: int = Field(ge=0)
    projects: int = Field(ge=0)
    activities: int = Field(ge=0)
    total: int = Field(ge=0)


class ReportSummaryRead(_StrictModel):
    local_date: date
    date_from: date | None
    date_until: date | None
    category_id: uuid.UUID | None
    responsible_user_id: uuid.UUID | None
    counts: ReportSummaryCounts


class ReportPeriod(_StrictModel):
    date_from: date | None
    date_until: date | None


class ReportCommonFilters(_StrictModel):
    category_id: uuid.UUID | None
    responsible_user_id: uuid.UUID | None


class TaskReportMetrics(_StrictModel):
    total_count: int = Field(ge=0)
    pending_count: int = Field(ge=0)
    completed_count: int = Field(ge=0)
    not_completed_count: int = Field(ge=0)
    resolved_count: int = Field(ge=0)
    completion_rate: Decimal | None = Field(default=None, ge=0, le=100)


class TaskReportGroup(TaskReportMetrics):
    key: str
    label: str


class TaskReportEvolution(TaskReportMetrics):
    planned_date: date


class TaskReportRead(_StrictModel):
    period: ReportPeriod
    filters: ReportCommonFilters
    master_task_id: uuid.UUID | None
    custom_tasks: bool | None
    summary: TaskReportMetrics
    by_task: list[TaskReportGroup]
    by_category: list[TaskReportGroup]
    evolution: list[TaskReportEvolution]


class ProgressReportMetrics(_StrictModel):
    total_count: int = Field(ge=0)
    no_iniciado_count: int = Field(ge=0)
    en_proceso_count: int = Field(ge=0)
    finalizado_count: int = Field(ge=0)
    configuracion_incompleta_count: int = Field(default=0, ge=0)
    average_progress: Decimal | None = Field(default=None, ge=0, le=100)


class ComplianceMetrics(_StrictModel):
    en_plazo_count: int = Field(ge=0)
    atrasado_count: int = Field(ge=0)
    con_adelanto_count: int = Field(ge=0)
    a_tiempo_count: int = Field(ge=0)
    con_retraso_count: int = Field(ge=0)


class ProgressCategoryGroup(ProgressReportMetrics):
    category_id: uuid.UUID
    category_name: str


class ProgressEvolution(_StrictModel):
    planned_date: date
    total_count: int = Field(ge=0)
    average_progress: Decimal | None = Field(default=None, ge=0, le=100)


class PendingItemReportRead(_StrictModel):
    period: ReportPeriod
    filters: ReportCommonFilters
    summary: ProgressReportMetrics
    compliance: ComplianceMetrics
    by_category: list[ProgressCategoryGroup]
    evolution: list[ProgressEvolution]


class ProjectReportRow(_StrictModel):
    project_id: uuid.UUID
    project_name: str
    category_id: uuid.UUID
    category_name: str
    planned_date: date | None
    progress: Decimal | None = Field(default=None, ge=0, le=100)
    state: str
    stage_count: int = Field(ge=0)


class ProjectReportRead(_StrictModel):
    period: ReportPeriod
    filters: ReportCommonFilters
    summary: ProgressReportMetrics
    stage_compliance: ComplianceMetrics
    by_category: list[ProgressCategoryGroup]
    by_project: list[ProjectReportRow]
    evolution: list[ProgressEvolution]
